Add Block Device Support (#191)
raspberrypi/picotool: README.md

BDEV:
    Commands related to embedded block devices

SYNOPSIS:
    picotool bdev ls <dirname> [-r] [-p <partition>] [--filesystem <fs>] [-f] [device-selection]
    picotool bdev mkdir <dirname> [-p <partition>] [--filesystem <fs>] [-f] [device-selection]
    picotool bdev cp <src> <dest> [-p <partition>] [--filesystem <fs>] [-f] [device-selection]
    picotool bdev rm <filename> [-p <partition>] [--filesystem <fs>] [-f] [device-selection]
    picotool bdev cat <filename> [-p <partition>] [--filesystem <fs>] [-f] [device-selection]

SUB COMMANDS:
    ls      List contents of the block device
    mkdir   Create directory on the block device
    cp      Copy file to/from the block device - use :filename to indicate files on the device (eg `cp main.py :main.py` to upload to the
            device)
    rm      Delete a file or an empty directory on the block device
    cat     Print contents of file on the block device
```

### ls

List contents of the block device

```text
$ picotool help bdev ls
BDEV LS:
    List contents of the block device

SYNOPSIS:
    picotool bdev ls <dirname> [-r] [-p <partition>] [--filesystem <fs>] [-f] [device-selection]

OPTIONS:
        <dirname>

>             The file name

Reviewer: "dirname" vs. "file name" ? :wink: 
Also, if I wanted to list the "root directory" would I give it a `<dirname>` of `/`, or something else?

Author: Just pass no `dirname`, like you would to the normal `ls` command - it’s an optional argument

        -r, --recursive
            List files in directories recursively
    Block device options
        -p, --partition
            Partition to use as block device
        <partition>
            partition number
        --filesystem
            Specify filesystem to use
        <fs>
            littlefs|fatfs

>         -f, --format

Reviewer: Hmmm, we now seem to have **BOTH** `-f, --format` and `-f, --force` flags?! Either this worn't work properly, or if it does work it's pretty confusing!

Author: Good spot - I’ll make this filesystem one not have a short argument

Reviewer: (perhaps the arg-parsing library could/should raise an error if given the same argument multiple times?)

Author: > (perhaps the arg-parsing library could/should raise an error if given the same argument multiple times?)

I will leave that for a future PR

    Target device selection
        --bus <bus>
            Filter devices by USB bus number
        --address <addr>
            Filter devices by USB device address
        --vid <vid>
            Filter by vendor id
        --pid <pid>
            Filter by product id
        --ser <ser>
            Filter by serial number
        -f, --force
            Force a device not in BOOTSEL mode but running compatible code to reset so the command can be executed. After executing the
            command (unless the command itself is a 'reboot') the device will be rebooted back to application mode
        -F, --force-no-reboot
            Force a device not in BOOTSEL mode but running compatible code to reset so the command can be executed. After executing the
            command (unless the command itself is a 'reboot') the device will be left connected and accessible to picotool, but without the
            USB drive mounted
```

### mkdir

Create directory on the block device

```text
$ picotool help bdev mkdir
BDEV MKDIR:
    Create directory on the block device

SYNOPSIS:
    picotool bdev mkdir <dirname> [-p <partition>] [--filesystem <fs>] [-f] [device-selection]

OPTIONS:

>         <dirname>
>             The file name

Reviewer: ```suggestion
        <dirname>
            The directory name
```

        <partition>
            partition number
        --filesystem
            Specify filesystem to use
        <fs>
            littlefs|fatfs
        -f, --format
            Format the drive if necessary (may result in data loss)
    Target device selection
        --bus <bus>
            Filter devices by USB bus number
        --address <addr>
            Filter devices by USB device address
        --vid <vid>
            Filter by vendor id
        --pid <pid>
            Filter by product id
        --ser <ser>
            Filter by serial number
        -f, --force
            Force a device not in BOOTSEL mode but running compatible code to reset so the command can be executed. After executing the
            command (unless the command itself is a 'reboot') the device will be rebooted back to application mode
        -F, --force-no-reboot
            Force a device not in BOOTSEL mode but running compatible code to reset so the command can be executed. After executing the
            command (unless the command itself is a 'reboot') the device will be left connected and accessible to picotool, but without the
            USB drive mounted
```

### cp

> Copy file to/from the block device - use :filename to indicate files on the device (eg `cp main.py :main.py` to upload to the device)

Reviewer: Probably doesn't hurt to be explicit?
```suggestion
Copy file to/from the block device - use :filename to indicate files on the device (e.g. `cp main.py :main.py` to upload to the device, or `cp :main.py main.py` to download from the device)
```

> Copy file to/from the block device - use :filename to indicate files on the device (eg `cp main.py :main.py` to upload to the device)

Reviewer: * Does this `bdev cp` command check that one (and only one) of either `src` or `dest` starts with a colon?
* Does it only work for files and not directories?
* If the answer to the previous question is "only files", would it be a lot of additional work to be able to copy directories? (which would presumably need to operate recursively)

Author: - No - you can copy a file to somewhere else on the device, or even use this to copy local files if you really wanted
- Yes, it only works for files
- Could probably add that - just needs a recursive walk like `ls` has

Author: Coming back to recursive copying, I think that is trickier, as it needs a recursive walk on the host side - I think I won't add that in this PR, but maybe a future `-r` argument could be added

        <partition>
            partition number
        --filesystem
            Specify filesystem to use
        <fs>
            littlefs|fatfs
        -f, --format
            Format the drive if necessary (may result in data loss)
    Target device selection
        --bus <bus>
            Filter devices by USB bus number
        --address <addr>
            Filter devices by USB device address
        --vid <vid>
            Filter by vendor id
        --pid <pid>
            Filter by product id
        --ser <ser>
            Filter by serial number
        -f, --force
            Force a device not in BOOTSEL mode but running compatible code to reset so the command can be executed. After executing the
            command (unless the command itself is a 'reboot') the device will be rebooted back to application mode
        -F, --force-no-reboot
            Force a device not in BOOTSEL mode but running compatible code to reset so the command can be executed. After executing the
            command (unless the command itself is a 'reboot') the device will be left connected and accessible to picotool, but without the
            USB drive mounted
```

### cat

> Print contents of file on the block device

Reviewer: Will an appropriate error get displayed if you try to `bdev cat` a directory name?

Author: I've pushed a change to print error strings instead of numbers, so you'd get `ERROR: LittleFS Open Error: Entry is a dir`

    picotool bdev rm <filename> [-p <partition>] [--filesystem <fs>] [--format] [device-selection]
    picotool bdev cat <filename> [-p <partition>] [--filesystem <fs>] [--format] [device-selection]

SUB COMMANDS:
    ls      List contents of the block device
    mkdir   Create directory on the block device
    cp      Copy file to/from the block device - use :filename to indicate files on the device (eg `cp main.py :main.py` to upload to the
            device)
    rm      Delete a file or an empty directory on the block device
    cat     Print contents of file on the block device
```

### ls

List contents of the block device

```text
$ picotool help bdev ls
BDEV LS:
    List contents of the block device

SYNOPSIS:
    picotool bdev ls <dirname> [-r] [-p <partition>] [--filesystem <fs>] [--format] [device-selection]

OPTIONS:
        <dirname>
            The directory name to list (optional)
        -r, --recursive
            List files in directories recursively
    Block device options
        -p, --partition
            Partition to use as block device
        <partition>
            partition number

>         --filesystem

Reviewer: Is the `--filesystem` flag only needed when formatting a block-device, or are there likely to be circumstances in which the filesystem auto-deteciton doesn't work?

Author: With the current supported filesystems it should only be needed when formatting, but future filesystems may be harder to auto-detect so may require this flag

            Specify filesystem to use
        <fs>
            littlefs|fatfs
        --format

>             Format the drive if necessary (may result in data loss)

Reviewer: Presumably using the `--format` option in conjunction with `picotool bdev ls` will **always** return an empty directory listing? :joy:  But this makes me wonder why `--format` is an option that can be used in combination  with any other `picotool bdev` command, rather than just being a standalone `picotool bdev format` or `picotool format_bdev` command? :thinking:

Author: I added it as a flag because of where the formatting occurs in the code - it will only format the drive if no filesystem (of the type specified by `--filesystem` if that is passed) is detected, so you could always call your first command with `--format` if you wanted to make sure you always do the command. Eg this works for both formatting new devices and updating the file on old devices:
```
picotool bdev cp main.py :/ --format --filesystem littlefs
```

Reviewer: Oh! I see now that I'd misread "Format the drive if necessary" as just "Format the drive" :facepalm: 

So I guess that means if you had a blockdevice that was already fatfs-formatted, and you wanted to explicitly reformat it, you'd need to do something like:
```
picotool bdev ls --format --filesystem littlefs
picotool bdev ls --format --filesystem fatfs
```
?

Author: Yes, so it probably is worth adding a separate `picotool bdev format` command to unconditionally format the block device

Author: Added

            Specify filesystem to use
        <fs>
            littlefs|fatfs
        --format
            Format the drive if necessary (may result in data loss)
    Target device selection
        --bus <bus>
            Filter devices by USB bus number
        --address <addr>
            Filter devices by USB device address
        --vid <vid>
            Filter by vendor id
        --pid <pid>
            Filter by product id
        --ser <ser>
            Filter by serial number
        -f, --force
            Force a device not in BOOTSEL mode but running compatible code to reset so the command can be executed. After executing the
            command (unless the command itself is a 'reboot') the device will be rebooted back to application mode
        -F, --force-no-reboot
            Force a device not in BOOTSEL mode but running compatible code to reset so the command can be executed. After executing the
            command (unless the command itself is a 'reboot') the device will be left connected and accessible to picotool, but without the
            USB drive mounted
```

### cat

Print contents of file on the block device

```text
$ picotool help bdev cat
BDEV CAT:

>     Print contents of file on the block device

Reviewer: So I guess `picotool bdev cat main.py > main.py` effectively does the same thing as `picotool bdev cp :main.py main.py` ?

Author: Yes - this also allows for things like `picotool bdev cat main.py | sha256sum`

            The file name
```

## bdev

> The `bdev` commands are for interacting with block devices in Flash. These can either be defined using binary info, or you can specify a partition to use as a block device. They work with a LittleFS filesystem, or a FatFS filesystem.

Reviewer 2: These can be defined is confusing; who is defining what, when? perhaps you mean selected not defined. Also perhaps "the commands" rather than They

Author: ```suggestion
The `bdev` commands are for interacting with block devices in Flash. The block device location can either be determined using binary info, or you can specify a partition to use as a block device. The commands work with a LittleFS filesystem, or a FatFS filesystem.
```

            Format the drive if necessary (may result in data loss)
    Target device selection
        --bus <bus>
            Filter devices by USB bus number
        --address <addr>
            Filter devices by USB device address
        --vid <vid>
            Filter by vendor id
        --pid <pid>
            Filter by product id
        --ser <ser>
            Filter by serial number
        --rp2040
            Assume the device is an RP2040 - this is only required when using a custom vid/pid with an RP2040 on Windows, and is ignored on
            other operating systems
        -f, --force
            Force a device not in BOOTSEL mode but running compatible code to reset so the command can be executed. After executing the
            command (unless the command itself is a 'reboot') the device will be rebooted back to application mode
        -F, --force-no-reboot
            Force a device not in BOOTSEL mode but running compatible code to reset so the command can be executed. After executing the
            command (unless the command itself is a 'reboot') the device will be left connected and accessible to picotool, but without the
            USB drive mounted

>         --bootsel-led <gpio>

Reviewer 2: probably a minor grumble, and maybe these are mostly applicable to all commands as you can force a reboot, and therefore there can be an activity LED, so maybe its fine... its just  abit weird that it is in target defice selection

Author: Yes, these are applicable to all commands that can force a reboot. Maybe the device selection could be refactored to show cleaner help text as it is the same for all commands (eg a `picotool help device_selection` command to list this help, rather than listing it out for all commands), but that would be for a different PR to do.

SYNOPSIS:
    picotool bdev ls <dirname> [-r] [-p <partition number>] [--partition-name <partition name>] [--partition-id <partition id>]

>                 [--filesystem <fs>] [--format] [--force-formattable] [--force-writeable] [device-selection]

Reviewer: Does the `bdev ls` command really need a `--force-writeable` command, given that `ls` will presumably never result in any writes?

Author: Technically yes, in order to allow formatting a read-only drive - not useful, but technically valid

Reviewer: Huh, so merely specifying `--force-formattable` wouldn't allow a read-only drive to be formatted? That might be surprising behaviour? :thinking:

Author: It throws a reasonable error - I think it's fine, as it requires you to stop and think even more about whether you should format the drive.
```
$ picotool bdev ls --partition-id 0 --format --force-formattable --filesystem littlefs
/
Formatting LittleFS file system
ERROR: This block device is not writeable
```

In this case, if you then pass `--force-writeable` too you still get an error, because the partition is only `NSBOOT(R)`
```
$ picotool bdev ls --partition-id 0 --format --force-formattable --force-writeable --filesystem littlefs
/
Formatting LittleFS file system
ERROR: The RP2350 device returned an error: permission failure
```

Reviewer: Hmmm, now that there's an explicit `bdev format` command (hooray), _maybe_ it doesn't make sense for the `bdev ls` command to have a `--format` flag, because presumably a freshly-formatted partition will always be empty? :thinking:

Author: It has uses - you might have a provisioning system that checks for the existence of a specifc file on a board, but which needs to deal with both formatted and un-formatted boards, and is expected to output formatted boards. You could use the same command for all of them because all `bdev` commands come with the ability to format first:
```
picotool bdev ls --format --filesystem littlefs
```

Reviewer: That seems a bit convoluted, but fair enough :slightly_smiling_face: 
If somebody wanted to do things "the long way round" i.e. avoiding any risk of data-loss, is there a `picotool bdev` command to check if a block device is correctly formatted or not?

Author: Yes, they can use any command except format, and the command will throw an error if the device wasn’t correctly formatted. If you pass `--filesystem` as well, it will throw an error if the device isn’t formatted to the specified file system.

Reviewer: Cool, thanks. Is that behaviour worth documenting (i.e. how to check if a block device is formatted, without writing to it), or do you think that it's sufficiently obvious that people will figure it out for themselves?

Author: I think it's clear enough?
```
$ picotool bdev ls
ERROR: No file system detected - to format the drive use `--format --filesystem <littlefs|fatfs>`

$ picotool bdev ls --filesystem littlefs
/
ERROR: LittleFS file system is corrupted - add --format flag to format it (this may result in data loss)

$ picotool bdev ls --filesystem fatfs
/
ERROR: FatFS file system is corrupted - add --format flag to format it (this may result in data loss)
```
